rocr: trap_handler_gfx12: restore wave sched mode

dayatsin-amd opened the pull request:
Store wave scheduler mode that was previously stored in ttmp1[26:27] on entry into ttmp11[26:27]. Then restore the value back to HW_REG_WAVE_SCHED_MODE on exit.

dayatsin-amd commented:
https://ontrack-internal.amd.com/browse/SWDEV-556301

lancesix commented on projects/rocr-runtime/runtime/hsa-runtime/core/runtime/trap_handler/trap_handler_gfx12.s:
```diff
@@ -200,14 +205,17 @@
 //   ttmp0: PC_LO[31:0] (Program Counter Low)
-//   ttmp1: PC_HI[15:0] (Program Counter High, bits 0-15), TrapID[3:0] (in bits 28-31 of original PC_HI)
+//   ttmp1: TrapID[3:0], SCHED_MODE[1:0], 0[9:0], PC_HI[15:0]
 //   ttmp11: 0[7:0], DebugEnabled[0], 0[15:0], NoScratch[0], 0[5:0]
```
While at it, could you update this line?  The top bits of ttmp11 are not guaranted to be 0-initialized, but are unspecified (this is why SCHED_MODE is passed in TTMP1 which has known values).

Also, NoScratch is a leftover from old architectures, there is no such bit used in gfx12.

Some of the bits there are preserved by the 1st level trap handler (for PC-parking), 

Should probably be something along the lines of:
```
// ttmp11: ?[7:0], DebugEnabled[0], PRESERVED[15:0], ?[6:0]
```

dayatsin-amd replied:
ACK

lancesix reviewed (approved):
Mostly one comment about a line you did not touch, otherwise this looks reasonable to me.

lancesix commented on projects/rocr-runtime/runtime/hsa-runtime/core/runtime/trap_handler/trap_handler_gfx12.s:
```diff
@@ -840,6 +848,10 @@
 .exit_trap:
+  // Restore ttmp11[27:26] into SCHED_MODE[0:1]
+  s_bfe_u32         ttmp2, ttmp11, TTMP11_SCHED_MODE_BFE
+  s_setreg_b32      hwreg(HW_REG_WAVE_SCHED_MODE), ttmp2
```
Actually, one thing we could improve is specify exactly what bits of SCHED MODE we are meaning to write.  We only read 2 bits out of ttmp11, so could be

```
s_setreg_b32      hwreg(HW_REG_WAVE_SCHED_MODE, 0, 2), ttmp2
```

dayatsin-amd replied:
ACK

fjankovi commented:
Does this change also need to check for KFD IOCTL version or it will work with 1.19 too? (1.20 adds trap handler support for exp. sched. mode in kfd: https://lore.kernel.org/amd-gfx/20251205140441.2094767-1-mario.limonciello@amd.com/)

cc: @superm1

lancesix commented:
> Does this change also need to check for KFD IOCTL version or it will work with 1.19 too? (1.20 adds trap handler support for exp. sched. mode in kfd: https://lore.kernel.org/amd-gfx/20251205140441.2094767-1-mario.limonciello@amd.com/)
> 
> cc: @superm1

No need to check the ioctl version.  We made sure to use bits which are guaranteed to be 0 initialized on "old" KFD versions, so we restore the "default" value in this case.

dayatsin-amd commented:
rebase the patch on top of  Vlad's latest changes.
Also added "if .amdgcn.gfx_generation_minor == 0" to make it easier to merge changes for future asics later on.

@lancesix  Can you please re-check

lancesix commented on projects/rocr-runtime/runtime/hsa-runtime/core/runtime/trap_handler/trap_handler_gfx12.s:
```diff
@@ -237,7 +251,7 @@
   s_cbranch_scc0    .handle_sw_trap                       // If not Stochastic, continue to check trap ID
 
-  s_load_b64           ttmp[14:15], ttmp[14:15], 0x8, scope:SCOPE_CU         // ttmp[14:15]=*stoch_trap_buf
+  s_load_b64        ttmp[14:15], ttmp[14:15], 0x8, scope:SCOPE_CU // ttmp[14:15]=*stoch_trap_buf
```
Unrelated, for easier tracking that should be a separate commit.

dayatsin-amd replied:
Fixes moved to this PR:
https://github.com/ROCm/rocm-systems/pull/3042

lancesix commented on projects/rocr-runtime/runtime/hsa-runtime/core/runtime/trap_handler/trap_handler_gfx12.s:
```diff
@@ -250,7 +264,7 @@
   // Branch if not a trap (an exception instead).
   s_bfe_u32         ttmp2, ttmp1, SQ_WAVE_PC_HI_TRAP_ID_BFE // ttmp2 = TrapID
-  s_cbranch_scc0       .check_exceptions			             // If TrapID is 0, it's an exception, so branch.
+  s_cbranch_scc0    .check_exceptions			                  // If TrapID is 0, it's an exception, so branch.
```
unrelated.

dayatsin-amd replied:
Fixes moved to this PR:
https://github.com/ROCm/rocm-systems/pull/3042

lancesix commented on projects/rocr-runtime/runtime/hsa-runtime/core/runtime/trap_handler/trap_handler_gfx12.s:
Should probably update this as well.

lancesix commented on projects/rocr-runtime/runtime/hsa-runtime/core/runtime/trap_handler/trap_handler_gfx12.s:
and here.

lancesix reviewed (commented):
Functionally this looks good, just some comments need updating and unrelated changes might be in their own patches (if you don't mind)

dayatsin-amd committed:
rocr: trap_handler_gfx12: restore wave sched mode

Store wave scheduler mode that was previously stored in ttmp1[26:27] on
entry into ttmp11[26:27]. Then restore the value back to
HW_REG_WAVE_SCHED_MODE on exit.

dayatsin-amd commented:
@lancesix  Updated patch to address comments:
https://github.com/ROCm/rocm-systems/compare/7110890d710fdae23fbef9d7b5c06034d50f8163..bf47c92fd28a7ce2940669de3d2385a7cce195c0

lancesix commented on projects/rocr-runtime/runtime/hsa-runtime/core/runtime/trap_handler/trap_handler_gfx12.s:
```diff
@@ -204,15 +213,20 @@
+    s_andn2_b32         ttmp11, ttmp11, TTMP11_SCHED_MODE_MASK
+    s_and_b32           ttmp2,  ttmp1, TTMP1_SCHED_MODE_MASK
+    s_or_b32            ttmp11, ttmp2, TTMP1_SCHED_MODE_MASK
+    s_mov_b32           ttmp3, 0
```
This `s_mov_b32 ttmp3, 0` should not be inside this `.if`.  It should be there regardless of the generation minor.

dayatsin-amd replied:
Oops. My bad.. nice catch! Fixed

lancesix reviewed (approved):
LGTM, thanks.
